### ISSUE TITLE
NHD via XML

### ISSUE THREAD

**Issue by yegor256** [yegor256/jpeek#93]
We had NHD metric implemented before in #3, but now our calculation mechanism is very different. We should re-implement OCC via XML/XSL.

**Comment by 0crat** [comment]
Job [#93](https://github.com/yegor256/jpeek/issues/93) is now in scope, role is `DEV`

**Comment by 0crat** [comment]
Bug was reported, see [§29](http://datum.zerocracy.com/pages/policy.html#29): +15 points just awarded to [@yegor256](http://www.0crat.com/u/yegor256), total is [+7255](http://www.0crat.com/u/yegor256)

**Comment by 0crat** [comment]
The job [#93](https://github.com/yegor256/jpeek/issues/93) assigned to icok. The budget is 30 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§8](http://datum.zerocracy.com/pages/policy.html#8) and [§9](http://datum.zerocracy.com/pages/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by iCok** [comment]
@0crat refuse

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/yegor256/jpeek/issues/93#issuecomment-357742651))

@iCok @icok[/z](http://www.0crat.com/u/icok) resigned from [#93](https://github.com/yegor256/jpeek/issues/93), please stop working

**Comment by 0crat** [comment]
Tasks refusal is discouraged, see [§6](http://datum.zerocracy.com/pages/policy.html#6): -15 points just awarded to @icok[/z](http://www.0crat.com/u/icok), total is [+15](http://www.0crat.com/u/icok)

**Comment by fabriciofx** [comment]
@0crat apply

**Comment by 0crat** [comment]
> @0crat apply ([here](https://github.com/yegor256/jpeek/issues/93#issuecomment-357766287))

@fabriciofx Can't understand "apply", try one of these:
  - `assign` Assign a performer to this issue
  - `hello` Just say hello
  - `in` Register this issue in scope (WBS)
  - `out` Remove this issue from scope (WBS)
  - `resign` Resign from this issue
  - `status` Check the status of the job
  - `version` Print current version of the bot
  - `waiting` Register an impediment for a job

**Comment by 0crat** [comment]
The job [#93](https://github.com/yegor256/jpeek/issues/93) assigned to @llorllale[/z](http://www.0crat.com/u/llorllale). The budget is 30 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§8](http://datum.zerocracy.com/pages/policy.html#8) and [§9](http://datum.zerocracy.com/pages/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by llorllale** [comment]
@yegor256 do you have the original PPT slides? Or at least a version that does not squash all layers of each single slide/page?

**Comment by yegor256** [comment]
@llorllale here you go (add it to our repo, instead of the slides):

[p123-counsell.pdf](https://github.com/yegor256/jpeek/files/1644102/p123-counsell.pdf)

**Comment by llorllale** [comment]
@yegor256 I'm blocked by issue #93, specifically the fact that there's currently no way to tell what `args` of a method are input parameters and which one is the result parameter:

From #93:
> our implementation counts not only parameter types of method but also it counts return types of method

From the paper you posted, *NHD* only deals with the method's argument types (see an example in page 129).

To illustrate the fact, here are a couple of examples (I've left out the `<ops>` tags):

A simple getter like `String id()`:

```xml
 <method abstract="false"
         ctor="false"
         desc="()Ljava/lang/String;"
         name="id"
         public="true"
         static="false">
    <args>
       <arg type="Ljava/lang/String">?</arg>
    </args>
 </method>
```

Here's another example produced from a method with signature `Optional<Issue> get(String id)`:

```xml
<method abstract="false"
        ctor="false"
        desc="(Ljava/lang/String;)Ljava/util/Optional;"
        name="get"
        public="true"
        static="false">
   <args>
      <arg type="Ljava/lang/String">?</arg>
      <arg type="Ljava/util/Optional">?</arg>
   </args>
</method>
```

I cannot tell if a method has a return type, and which is it.

**Comment by llorllale** [comment]
@0crat waiting

**Comment by llorllale** [comment]
Come to think of if, the `method/@desc` seems to be enclosing the input parameter types in parenthesis. Is this so? 

Still, having to parse @desc would be an ugly hack.

**Comment by 0crat** [comment]
> @0crat waiting ([here](https://github.com/yegor256/jpeek/issues/93#issuecomment-359050643))

@llorllale The impediment was registered successfully

**Comment by yegor256** [comment]
@llorllale #97 is fixed, I believe you can continue here

**Comment by llorllale** [comment]
blocked by #102

**Comment by llorllale** [comment]
@0crat waiting

**Comment by 0crat** [comment]
> @0crat waiting ([here](https://github.com/yegor256/jpeek/issues/93#issuecomment-359280609))

@llorllale Job [#93](https://github.com/yegor256/jpeek/issues/93) is already on hold

**Comment by 0crat** [comment]
@llorllale[/z](https://www.0crat.com/u/llorllale) this job was assigned to you 5 days ago. It will be taken away from you soon, unless you close it, see [§8](http://datum.zerocracy.com/pages/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by llorllale** [comment]
blocked by #103

**Comment by llorllale** [comment]
@0crat waiting

**Comment by 0crat** [comment]
> @0crat waiting ([here](https://github.com/yegor256/jpeek/issues/93#issuecomment-359832882))

@llorllale Job [#93](https://github.com/yegor256/jpeek/issues/93) is already on hold

**Comment by llorllale** [comment]
blocked as well  by #101

**Comment by 0crat** [comment]
@llorllale[/z](https://www.0crat.com/u/llorllale) this job was assigned to you 8 days ago. It will be taken away from you soon, unless you close it, see [§8](http://datum.zerocracy.com/pages/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by llorllale** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/jpeek/issues/93#issuecomment-360458846))

@llorllale This is what I know about this job, as in [§32](http://datum.zerocracy.com/pages/policy.html#32):

  * The job [#93](https://github.com/yegor256/jpeek/issues/93) is in scope for 11days
  * The role is `DEV`
  * The job is assigned to @llorllale[/z](https://www.0crat.com/u/llorllale) for 9days
  * These users are banned: @yegor256[/z](https://www.0crat.com/u/yegor256): This user reported the ticket; @icok[/z](https://www.0crat.com/u/icok): User was resigned from the ticket
  * The job has an impediment

**Comment by llorllale** [comment]
@yegor256 can we close this?

**Comment by yegor256** [comment]
@llorllale thanks!

**Comment by 0crat** [comment]
Order was successfully finished: +30 points just awarded to @llorllale[/z](https://www.0crat.com/u/llorllale), total is [+315](https://www.0crat.com/u/llorllale)

**Comment by 0crat** [comment]
The job [#93](https://github.com/yegor256/jpeek/issues/93) is now out of scope

**Comment by 0crat** [comment]
Job [#93](https://github.com/yegor256/jpeek/issues/93) was not in scope

**Comment by 0pdd** [comment]
@yegor256 the puzzle [#121](https://github.com/yegor256/jpeek/issues/121) is still not solved; solved: [#122](https://github.com/yegor256/jpeek/issues/122).
